src/cluster/sharding.rst

```diff
@@ -12,290 +12,490 @@
+::
 
-Views
-=====
+    [cluster]
+    placement = <zone-name-1>:2,<zone-name-2>:1
 
-The views need to be moved together with the shards. If you do not, then
-CouchDB will rebuild them and this will take time if you have a lot of
-documents.
+In this example, CouchDB will ensure that two replicas for a shard will
+be hosted on nodes with the zone attribute set to ``<zone-name-1>`` and
+one replica will be hosted on a new with the zone attribute set to
+``<zone-name-2>``.
 
-The views are stored in ``data/.shards``.
+This approach is flexible, since you can also specify zones on a per-
+database basis by specifying the placement setting as a query parameter
+when the database is created, using the same syntax as the ini file:
 
-It is possible to not move the views and let CouchDB rebuild the view every
-time you move a shard. As this can take quite some time, it is not recommended.
+.. code:: bash
 
-.. _cluster/sharding/preshard:
+    curl -X PUT $COUCH_URL:5984/<dbname>?zone=<zone>
 
-Reshard? No, Preshard!
-======================
+Note that you can also use this system to ensure certain nodes in the
+cluster do not host any replicas for newly created databases, by giving
+them a zone attribute that does not appear in the ``[cluster]``
+placement string.
 
-Reshard? Nope. It cannot be done. So do not create databases with too few
-shards.
+Resharding a database to a new q value
+--------------------------------------
 
-If you can not scale out more because you set the number of shards too low, then
-you need to create a new cluster and migrate over.
+The q value for a database can only be set when the database is created,
+precluding live resharding. Instead, to reshard a database, it must be
+regenerated. Here are the steps:
 
-#. Build a cluster with enough nodes to handle one copy of your data.
-#. Create a database with the same name, n=1 and with enough shards so you do
-   not have to do this again.
-#. Set up 2 way replication between the 2 clusters.
-#. Let it sync.
-#. Tell clients to use both the clusters.
-#. Add some nodes to the new cluster and add them as replicas.
-#. Remove some nodes from the old cluster.
-#. Repeat 6 and 7 until you have enough nodes in the new cluster to have 3
-   replicas of every shard.
-#. Redirect all clients to the new cluster
-#. Turn off the 2 way replication between the clusters.
-#. Shut down the old cluster and add the servers as new nodes to the new
-   cluster.
-#. Relax!
+1. Create a temporary database with the desired shard settings, by
+   specifying the q value as a query parameter during the PUT
+   operation.
+2. Replicate the primary database to the temporary one. Multiple
+   replications may be required if the primary database is under
+   active use.
+3. Delete the primary database. **Make sure nobody is using it!**
+4. Recreate the primary database with the desired shard settings.
+5. Replicate the temporary back to the primary.
+6. Delete the temporary database.
```

Comment:
This more describes an "offline" resharding. Rather than having the note in step 3, to be safe, I think there should be extra instructions:

1. Prior to (2), "stop clients accessing the database".
2. After (5), "clients can now access the database again".